File: typescript/packages/core/src/core/wandb/metadata.ts
import type { Run, RunVariables } from './types.ts'

export function runMetadata(run: Run, variables: RunVariables): Record<string, unknown> {
  const raw = run.systemMetrics
  const systemMetrics: unknown = typeof raw === 'string' ? JSON.parse(raw) : (raw ?? null)
  return {
    ...variables,
    path: [variables.entity, variables.project, variables.run].join('/'),
    storage_id: run.id ?? null,
    display_name: run.displayName ?? null,
    state: run.state ?? null,
    tags: run.tags ?? null,
    sweep_name: run.sweepName ?? null,
    group: run.group ?? null,
    job_type: run.jobType ?? null,
    commit: run.commit ?? null,
    read_only: run.readOnly ?? null,
    created_at: run.createdAt ?? null,
    heartbeat_at: run.heartbeatAt ?? null,
    description: run.description ?? null,
    notes: run.notes ?? null,
    user: run.user ?? null,
    system_metrics: systemMetrics,
    history_line_count: run.historyLineCount ?? null,
    history_keys: run.historyKeys ?? null,
    file_count: run.fileCount ?? null,
  }
}

File: python/mirage/core/wandb/metadata.py
import json
from typing import Any

from mirage.core.wandb.types import Run, RunVariables


def run_metadata(run: Run, variables: RunVariables) -> dict[str, Any]:
    system_metrics = run.get("systemMetrics")
    if isinstance(system_metrics, str):
        system_metrics = json.loads(system_metrics)
    run_path = "/".join(
        (variables["entity"], variables["project"], variables["run"]))
    return {
        **variables,
        "path": run_path,
        "storage_id": run.get("id"),
        "display_name": run.get("displayName"),
        "state": run.get("state"),
        "tags": run.get("tags"),
        "sweep_name": run.get("sweepName"),
        "group": run.get("group"),
        "job_type": run.get("jobType"),
        "commit": run.get("commit"),
        "read_only": run.get("readOnly"),
        "created_at": run.get("createdAt"),
        "heartbeat_at": run.get("heartbeatAt"),
        "description": run.get("description"),
        "notes": run.get("notes"),
        "user": run.get("user"),
        "system_metrics": system_metrics,
        "history_line_count": run.get("historyLineCount"),
        "history_keys": run.get("historyKeys"),
        "file_count": run.get("fileCount"),
    }
